### PR TITLE
[MRG][FIX] Fix GDF returning all annotations with same description

### LINKED ISSUE #5863: GDF events
Check this MWE. `mne.events_from_annotations` is not able to pick up the different events and they all get mapped to `'GDF event'`. Plus I'm not sure if we should expose `mne.io.find_edf_events` nor `edf_raw.find_edf_events()` because that would now work with `raw.find_edf_events()`. (and I would expect the same methods, no matter which reader did I use to load `raw`)

MWE
```py
"""
Disclaimer this MWE uses data from http://www.bbci.de/competition/iv/#dataset2a
"""
import mne
import numpy as np
from os import path as op

tmpdir = mne.utils._TempDir()
DROPBOX_FILE_URL = "https://www.dropbox.com/s/nn6dzc5cc7riz7c/A01E.gdf?dl=1"
fname = op.join(tmpdir, 'A01E.gdf')

mne.utils._fetch_file(url=DROPBOX_FILE_URL, file_name=fname,
                      hash_='ec3de227cd3bf1760accca0e3269590e')

raw = mne.io.read_raw_edf(fname)

# Regular workflow
events_A, event_id_A = mne.events_from_annotations(raw)
print(raw.annotations)
print(events_A.shape)
print(event_id_A)

# whatever
def my_events_from_annotations(edf_raw):
    _, onset, label, duration, _ = edf_raw.find_edf_events()
    events = np.stack([onset, duration, label], axis=1)
    event_id = {'GDF event {}'.format(chr(65 + idx)): l
                for idx, l in enumerate(set(label))}
    return events, event_id

events_B, event_id_B = my_events_from_annotations(raw)

print(events_B.shape)
print(event_id_B)

np.testing.assert_array_equal(events_A[:, 0], events_B[:, 0])
```
output
```<Annotations  |  595 segments : GDF event (595), orig_time : 2005-01-19 12:00:00>
(595, 3)
{'GDF event': 1}
(595, 3)
{'GDF event A': 768, 'GDF event B': 783, 'GDF event C': 1072, 'GDF event D': 276, 'GDF event E': 277, 'GDF event F': 32766, 'GDF event G': 1023}
```

ping: @jaeilepp

### PR DESCRIPTION
**Pull request by massich**
The snipped in #5863 now returns this:

``` shell
<Annotations  |  595 segments : 32766 (9), 276 (1), 277 (1)..., orig_time : 2005-01-19 12:00:00>
(595, 3)
{'32766': 1, '276': 2, '277': 3, '1072': 4, '768': 5, '783': 6, '1023': 7}
(595, 3)
{'GDF event A': 768, 'GDF event B': 783, 'GDF event C': 1072, 'GDF event D': 276, 'GDF event E': 277, 'GDF event F': 32766, 'GDF event G': 1023}
```

I'm not sure if the ids, are stored somewhere in the meatadata (ie: `4242` meaning something like `Start recording`)

fixes #5863

### CODE REVIEW

**Comment by massich** [comment]
(I'll deprecate `find_edf_events` in this PR)

**Comment by massich** [comment]
Just to keep my thoughts somewhere. 

The code at this point loads only the `gdf_events_lut` once from the file using global. The other option would be to use a singleton to avoid the global, or a lazy loader. 

I like this singleton implementation as a decorator from [here](https://riptutorial.com/python/example/10954/create-singleton-class-with-a-decorator) (I've modified the code here without testing anything, jsut as type-thinking)
```py
def singleton(cls):    
    instance = [None]
    def wrapper(*args, **kwargs):
        if instance[0] is None:
            instance[0] = cls(*args, **kwargs)
        return instance[0]

    return wrapper

@singleton
class SomeSingletonClass:
    LUT = None
    x = 2
    def __init__(self):
        print("Created!")
    
    def __call__(self):
        if LUT is not None:
              return LUT
        else:
              self.LUT = _get_gdf_events_lut_bla_bla('watever')
              return LUT
        

instance = SomeSingletonClass()  # prints: Created!
instance = SomeSingletonClass()  # doesn't print anything
print(instance.x)                # 2

print(SomeSingletonClass().x)    # 3

my_lut = instance()
```

But I think I'll just load it everytime the module is loaded. I don't think this is any kind of overhead for MNE.

**Comment by cbrnr** [comment]
I think you can load the file every time the GDF reader imports a file. I'm not 100% sure if keeping this in a file is the best idea, because this makes it easy for people to manipulate that file. We do have a file where we define various FIFF constants - why not do the same thing for GDF?

**Comment by massich** [comment]
The MWE works
```py
"""
Disclaimer this MWE uses data from http://www.bbci.de/competition/iv/#dataset2a
"""
import mne
import numpy as np
from os import path as op

tmpdir = mne.utils._TempDir()
DROPBOX_FILE_URL = "https://www.dropbox.com/s/nn6dzc5cc7riz7c/A01E.gdf?dl=1"
fname = op.join(tmpdir, 'A01E.gdf')

mne.utils._fetch_file(url=DROPBOX_FILE_URL, file_name=fname,
                      hash_='ec3de227cd3bf1760accca0e3269590e')

raw = mne.io.read_raw_edf(fname)

# Regular workflow
events, event_id = mne.events_from_annotations(raw)
print(raw.annotations)
print(events.shape)
print(event_id)
```
```
<Annotations  |  595 segments : start of a new segment (after a break) (9), eeg:Idling EEG - eyes open (1), eeg:Idling EEG - eyes closed (1)..., orig_time : 2005-01-19 12:00:00>
(595, 3)
{'start of a new segment (after a break)': 1, 'eeg:Idling EEG - eyes open': 2, 'eeg:Idling EEG - eyes closed': 3, 'Unknown': 4, 'Start of Trial, Trigger at t=0s': 5, 'cue unknown/undefined (used for BCI competition) ': 6, 'Rejection of whole trial': 7}
```

But there's something wrong with the gdf tests. (not the missing file, I'll amend that)

File: mne/io/edf/tests/test_edf.py [hunk 1]
```diff
@@ -197,6 +198,16 @@ def test_to_data_frame():
         assert_array_equal(df.values[:, 0], raw._data[0] * 1e13)
 
 
+def test_find_edf_events_depracation():
```
Comment (agramfort) [inline mne/io/edf/tests/test_edf.py]: depracation -> deprecation

File: mne/io/edf/tests/test_gdf.py [hunk 2]
```diff
@@ -77,4 +79,9 @@ def test_gdf2_data():
                      eog=None, misc=None)
 
 
+def test_gdf_events_lut():
+    """Test something about GDF_EVENTS_LUT to make sure it has not change."""
```
Comment (agramfort) [inline mne/io/edf/tests/test_gdf.py]: change -> changed

File: mne/io/edf/edf.py [hunk 3]
```diff
@@ -1223,3 +1219,20 @@ def _get_edf_default_event_id(descriptions):
...
+    events = edf_info.get('events', None)
+    # Annotations in GDF: events are stored as the following
+    # list: `events = [n_events, pos, typ, chn, dur]` where pos is the
+    # latency, dur is the duration in samples. They both are
+    # numpy.ndarray
+    if events is not None and events[1].shape[0] > 0:
+        onset = events[1] / sfreq
+        duration = events[4] / sfreq
+        desc = [GDF_EVENTS_LUT[key]
+                if key in GDF_EVENTS_LUT.keys() else 'Unknown'
```
Comment (agramfort) [inline mne/io/edf/edf.py]: GDF_EVENTS_LUT.keys() -> GDF_EVENTS_LUT

if you put the .keys() the "in" cost you a O(n) vs O(1) (list vs hash table)

**Comment by agramfort** [comment]
@massich besides my nitpicks and red CIs diff looks good to me.

File: mne/io/edf/_utils.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#          Joan Massich <mailsik@gmail.com>
+#
+# License: BSD (3-clause)
+
+import re
+from ...utils import hashfunc
+
+
+def _load_gdf_events_lut(fname, md5):
+    assert hashfunc(fname, hash_type='md5') == md5
```
Comment (cbrnr) [inline mne/io/edf/_utils.py]: My usual complaint, since this is not a test file I would use try/except instead of `assert`.

**Comment by agramfort** [comment]
green.

thx @massich

**Comment by massich** [comment]
Thx for the last mile.

**Pull request merged**